BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Authentication/CredentialEnvelope.cs

```diff
@@ -0,0 +1,41 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Authentication
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// An envelope for an instance of Azure.Core.TokenCredential.
+    /// </summary>
+    public abstract class CredentialEnvelope
+    {
+        /// <summary>
+        /// Source: (https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-acquire-cache-tokens#scopes-when-acquiring-tokens).
+        /// </summary>
+        private const string Scope = "https://monitor.azure.com"; // TODO: THIS SCOPE IS UNVERIFIED. WAITING FOR SERVICES TEAM TO PROVIDE AN INT ENVIRONMENT FOR E2E TESTING.
+
+        /// <summary>
+        /// Gets the TokenCredential object held by this class.
+        /// </summary>
+        public abstract object Credential { get; }
+
+        /// <summary>
+        /// Gets an Azure.Core.AccessToken.
+        /// </summary>
+        /// <param name="cancellationToken">The System.Threading.CancellationToken to use.</param>
+        /// <returns>A valid Azure.Core.AccessToken.</returns>
+        public abstract string GetToken(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Gets an Azure.Core.AccessToken.
+        /// </summary>
+        /// <param name="cancellationToken">The System.Threading.CancellationToken to use.</param>
+        /// <returns>A valid Azure.Core.AccessToken.</returns>
+        public abstract Task<string> GetTokenAsync(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Get scopes for Azure Monitor as an array.
+        /// </summary>
+        /// <returns>An array of scopes.</returns>
+        protected static string[] GetScopes() => new string[] { Scope };
```

Comment:
PERF: This creates a new array each time it's called. Consider using a static field -- since it's a compile-time constant.

Reply:
Ah, I see that derived classes call it just once and cache the result in a TokenRequestContext.
Note that TokenRequestContext also includes a "ParentRequestId" property -- currently not being set or used here. You may find it useful to set that ID so that you have a correlation ID for AAD failures. At that point, you'll want to create a new TokenRequestContext object each time.